src/os2.rs

```diff
@@ -297,3 +299,139 @@ deserialize_visitor!(
         Ok(res)
     }
 );
+
+impl os2 {
+    /// Populate ulCodePageRange fields using a
+    pub fn int_list_to_code_page_ranges(&mut self, bitlist: &Vec<u8>) {
+        let mut code_pages1 = bitlist.clone();
+        code_pages1.sort_unstable();
+        let split_at = code_pages1
+            .iter()
+            .position(|&x| x >= 32)
+            .unwrap_or_else(|| bitlist.len());
+        let mut code_pages2 = code_pages1.split_off(split_at);
+        code_pages2.iter_mut().for_each(|x| *x -= 32);
+
+        self.ulCodePageRange1 = Some(int_list_to_num(&code_pages1) as u32);
+        self.ulCodePageRange2 = Some(int_list_to_num(&code_pages2) as u32);
+    }
+    /// implementation based on ufo2ft:
+    /// https://github.com/googlefonts/ufo2ft/blob/main/lib/ufo2ft/util.py#l307
+    pub fn calc_code_page_ranges(&mut self, mapping: &BTreeMap<u32, u16>) {
```

Comment:
So now that I have a clearer idea of what we're doing, it feels like what we really want here is a function that creates two 32-bit integers in place. This is the sort of low-level thing that can often be annoying to do, but with Rust we could make this quite easy. What I would personally do is make a helper type that looks like a collection but is just doing bit arithmetic under the hood, something like:


```rust
#[derive(Default)]
struct CodePageRanges {
    one: u32,
    two: u32,
}

impl CodePageRanges {
    fn set(&mut self, idx: usize) {
        assert!(idx < 64);
        if idx >= 32 {
            self.two |= 1 << (idx - 32);
        } else {
            self.one |= 1 << idx;
        }
    }
    
    fn into_inner(self) -> (u32, u32) {
        (self.one, self.two)
    }
}
```

And then you can just call this directly, everwhere that you are using `code_page_ranges`:

```rust
let mut code_page_ranges = CodePageRanges::default();

if unicodes_contains('Þ') && has_ascii {
    code_page_ranges.set(0); // Latin 1
}
```

This then does everything at once; we're building up our final representation directly, as we add each new value.

Reply:
Nice! this is much cleaner.